### PR TITLE
Lpfg 576 event date and time

### PR DESCRIPTION
**Pull request by Peter-D-Taylor**
This needs to be merged along side Lpfg 576 event date and time in lpg-learning-catalouge

Created eventController.ts
Added eventController and eventValidator to applicationContext.ts
Added create, get, update functions to index.ts
Changed date to dateRanges in event.ts
Added get and set eventFactory functions to validator.ts
Added date validation and parsing to datetime.ts
Added error messages to en.json
Added eventController routes to server.ts
Created eventControllerTest.ts
Created _events.scss
Created _event-preview.scss
Create errorSumary template.njk and macro.njk
Create timeInputs template.njk and macro.njk
Created events-preview.html
Create events.html

### CODE REVIEW

File: src/controllers/module/event/eventController.ts [hunk 1]
```diff
@@ -21,41 +21,105 @@ export class eventController {
 	}
 
 	private setRouterPaths() {
+		this.router.param('courseId', async (req, res, next, courseId) => {
```
Comment (jakub-kainos) [inline src/controllers/module/event/eventController.ts]: worth extracting to a separate method like eg. `setParameterMappingFunctions`

File: src/learning-catalogue/model/event.ts [hunk 2]
```diff
@@ -1,17 +1,13 @@
...
-		message: 'validation.module.event.date.empty',
+		groups: ['all', 'event.all', 'event.dateRanges'],
+		message: 'validation.module.event.dateRanges.empty',
 	})
-	@MinDate(new Date(Date.now()), {
-		groups: ['all', 'event.all', 'event.date'],
-		message: 'validation.module.event.date.past',
-	})
-	date: Date
+	dateRanges: any
```
Comment (jakub-kainos) [inline src/learning-catalogue/model/event.ts]: would be good to introduce a strong type, like `Array<DateRange>`

File: src/learning-catalogue/validator/validator.ts [hunk 3]
```diff
@@ -34,6 +34,14 @@ export class Validator<T> {
 		this._factory = value
 	}
 
+	get eventFactory() {
```
Comment (jakub-kainos) [inline src/learning-catalogue/validator/validator.ts]: why do we need a getter and setter here? Factory is passed as a constructor parameter - do we ever need to retrieve it later?

File: src/lib/datetime.ts [hunk 4]
```diff
@@ -1,7 +1,11 @@
+var isValidDate = require('is-valid-date')
```
Comment (jakub-kainos) [inline src/lib/datetime.ts]: `const` instead of `var`

File: src/learning-catalogue/model/event.ts [hunk 5]
```diff
@@ -1,17 +1,13 @@
...
+import {IsNotEmpty, IsPositive} from 'class-validator'
 
 export class Event {
 	id: string
 
 	@IsNotEmpty({
-		groups: ['all', 'event.all', 'event.date'],
-		message: 'validation.module.event.date.empty',
+		groups: ['all', 'event.all', 'event.dateRanges'],
+		message: 'validation.module.event.dateRanges.empty',
```
Comment (m24murray) [inline src/learning-catalogue/model/event.ts]: I know we have a bit of both in the app right now - but id prefer if we were using underscores to delimit our keys

File: test/unit/controllers/module/moduleControllerTest.ts [hunk 6]
```diff
@@ -67,9 +65,7 @@ describe('Module Controller Tests', function() {
 		await setModule(request, response)
 
 		//To be done - would expect to render form for specific module type
```
Comment (m24murray) [inline test/unit/controllers/module/moduleControllerTest.ts]: Can we get rid of this comment?

File: src/locale/en.json [hunk 7]
```diff
@@ -9,6 +9,7 @@
 	"course_title_add_title": "Add course title",
 	"course_overview_title": "Course overview",
 	"course_preview_title": "Course preview",
+	"event_add": "Add event date & time",
```
Comment (m24murray) [inline src/locale/en.json]: append with '_title' to make consistent with with other titles.

**Pull request merged**